### PR TITLE
3923- url creation with optional port flag

### LINKED ISSUE #3923: `odo url create` shouldn't require a port if only one port exists in the devfile
/kind bug
/area devfile

## What versions of software are you using?

**Operating System:**
macOS

**Output of `odo version`:**
master

## How did you run odo exactly?
`odo url create --host <ingress-domain>`

## Actual behavior
`odo url create --host <ingress-domain>` without the `--port` flag fails:
```
Johns-MacBook-Pro-3:dev2 johncollier$ odo create --devfile devfile.yaml 
Experimental mode is enabled, use at your own risk

Validation
 ✓  Creating a devfile component from devfile path: /Users/johncollier/ododev/dev2/devfile.yaml [106576ns]
 ✓  Validating devfile component [206951ns]

Please use `odo push` command to create the component with source deployed
Johns-MacBook-Pro-3:dev2 johncollier$ odo url create --host johncollier.ca
 ✗  port must be provided to create an endpoint entry in devfile
```

## Expected behavior
odo detects that there's only one port in the devfile and saves me from manually having to look up the port in the devfile. I believe this was the behaviour in past odo versions.

## Any logs, error output, etc?
Sample devfile:
```
schemaVersion: 2.0.0
metadata:
  name: nodejs
  version: 1.0.0
  alpha.build-dockerfile: "https://raw.githubusercontent.com/odo-devfiles/registry/master/devfiles/nodejs/build/Dockerfile"
  alpha.deployment-manifest: "https://raw.githubusercontent.com/odo-devfiles/registry/master/devfiles/nodejs/deploy/deployment-manifest.yaml"
starterProjects:
  - name: nodejs-starter
    git:
      remotes:
        origin: "https://github.com/odo-devfiles/nodejs-ex.git"
components:
  - container:
      name: runtime
      image: registry.access.redhat.com/ubi8/nodejs-12:1-45
      memoryLimit: 1024Mi
      mountSources: true
      sourceMapping: /project
      endpoints:
        - name: http-3000
          targetPort: 3000
commands:
  - exec:
      id: install
      component: runtime
      commandLine: npm install
      workingDir: /project
      group:
        kind: build
        isDefault: true
  - exec:
      id: run
      component: runtime
      commandLine: npm start
      workingDir: /project
      group:
        kind: run
        isDefault: true
  - exec:
      id: debug
      component: runtime
      commandLine: npm run debug
      workingDir: /project
      group:
        kind: debug
        isDefault: true
  - exec:
      id: test
      component: runtime
      commandLine: npm test
      workingDir: /project
      group:
        kind: test
        isDefault: true
```

### PR DESCRIPTION
**Pull request by yangcao77**
**What type of PR is this?**

/kind bug
/area devfile
/area url

**What does does this PR do / why we need it**:
The `--port` flag should be an optional flag if there is only one port exposed in devfile. 
This PR fixes the regression.

Also created an integration test for the specific case

**Which issue(s) this PR fixes**:

Fixes #3923 

**PR acceptance criteria**:

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [ ] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
`make test-cmd-devfile-url`

Manual Test: 
`odo url create` should successfully create an endpoint in devfile

### CODE REVIEW

File: tests/integration/devfile/cmd_devfile_url_test.go [hunk 1]
```diff
@@ -115,6 +115,18 @@ var _ = Describe("odo devfile url command tests", func() {
...
+		It("should create a URL without port flag if only one port exposed in devfile", func() {
+			url1 := helper.RandString(5)
+			host := helper.RandString(5) + ".com"
+
+			helper.CmdShouldPass("odo", "create", "nodejs", "--project", namespace, componentName)
+
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(context, "devfile.yaml"))
+
+			helper.CmdShouldPass("odo", "url", "create", url1, "--host", host, "--secure", "--ingress")
```
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_url_test.go:127]: please add a check to verify that the default port was added to the devfile.

Comment (yangcao77) [reply tests/integration/devfile/cmd_devfile_url_test.go]: updated

File: tests/integration/devfile/cmd_devfile_url_test.go [hunk 2]
```diff
@@ -115,6 +115,20 @@ var _ = Describe("odo devfile url command tests", func() {
...
+			host := helper.RandString(5) + ".com"
+
+			helper.CmdShouldPass("odo", "create", "nodejs", "--project", namespace, componentName)
+
+			helper.CopyExample(filepath.Join("source", "devfiles", "nodejs", "project"), context)
+			helper.CopyExampleDevFile(filepath.Join("source", "devfiles", "nodejs", "devfile.yaml"), filepath.Join(context, "devfile.yaml"))
+
+			helper.CmdShouldPass("odo", "url", "create", url1, "--host", host, "--ingress")
+			stdout := helper.CmdShouldPass("odo", "url", "list")
+			helper.MatchAllInOutput(stdout, []string{url1, "Not Pushed"})
```
Comment (mik-dass) [inline tests/integration/devfile/cmd_devfile_url_test.go:129]: Please also check the port number in the output

**Review by mik-dass — APPROVED** [review]
/lgtm

**Comment by mik-dass** [comment]
```
error: some steps failed:
  * could not run steps: execution cancelled
time="2020-09-16T17:52:27Z" level=info msg="Reporting job state 'failed' with reason 'executing_graph:interrupted'"
{"component":"entrypoint","file":"prow/entrypoint/run.go:247","func":"k8s.io/test-infra/prow/entrypoint.gracefullyTerminate","level":"error","msg":"Process gracefully exited before 30m0s grace period","severity":"error","time":"2020-09-16T17:52:27Z"}
```
/retest

**Pull request merged**